### PR TITLE
sync_to_1.1: avoiding load object meta when object stats are valid in `mo_table_size` and `metadata_scan`

### PR DESCRIPTION
**Pull request by gouhongshen**
## What type of PR is this?

- [ ] API-change
- [x] BUG
- [x] Improvement
- [ ] Documentation
- [ ] Feature
- [ ] Test and CI
- [ ] Code Refactoring

## Which issue(s) this PR fixes:

issue #13013 #13151 

## What this PR does / why we need it:
1. adding debug log to dump catalog cache  when `ExpectedEOB` for #13151 
2. avoid loading object meta when object stats are valid when `mo_table_size` and `metadata_scan`.

### CODE REVIEW

**Comment by matrix-meow** [comment]
@gouhongshen Thanks for your contributions!

**Here are review comments for file pkg/sql/plan/function/func_mo.go:**
Title: sync_to_1.1: avoiding load object meta when object stats are valid in `mo_table_size` and `metadata_scan`

Body: This pull request addresses two issues (#13013 and #13151) by adding debug logs to dump catalog cache and avoiding loading object metadata when object stats are valid in the `mo_table_size` and `metadata_scan` functions.

Changes made in `pkg/sql/plan/function/func_mo.go`:
- Added a check for `moerr.OkExpectedEOB` error code in the `MoTableRows` function. If this error is encountered, a debug log is printed and an `InvalidArgNoCtx` error is returned.
- Added a check for `moerr.OkExpectedEOB` error code in the `MoTableSize` function. If this error is encountered, a debug log is printed and an `InvalidArgNoCtx` error is returned.

Review:

1. The title of the pull request is clear and concise, indicating that the changes aim to avoid loading object metadata when object stats are valid in specific functions.

2. The body of the pull request provides information about the type of PR (bug fix and improvement) and the issues it addresses (#13013 and #13151). It also explains the purpose of the changes, which is to add debug logs and avoid loading object metadata when not necessary.

3. The changes in `pkg/sql/plan/function/func_mo.go` include the addition of a check for the `moerr.OkExpectedEOB` error code in both the `MoTableRows` and `MoTableSize` functions. If this error is encountered, a debug log is printed and an `InvalidArgNoCtx` error is returned.

Suggestions for improvement:

1. It would be helpful to provide more context about the issues being addressed (#13013 and #13151) in the body of the pull request. This could include a brief description of the issues and why they need to be fixed.

2. The debug logs added in the `MoTableRows` and `MoTableSize` functions are currently commented as "TODO(ghs)" and "is debug for #13151, will remove later" respectively. It would be better to provide more meaningful comments explaining the purpose of these debug logs.

3. Consider providing more detailed commit messages for the changes made in `pkg/sql/plan/function/func_mo.go`. This would make it easier for future developers to understand the changes and their purpose.

4. It's unclear why the `InvalidArgNoCtx` error is returned when encountering the `moerr.OkExpectedEOB` error code. It would be helpful to provide a comment or documentation explaining the reason for this specific error handling.

5. Consider optimizing the code by consolidating the error handling logic for the `moerr.OkExpectedEOB` error code. Instead of duplicating the code in both the `MoTableRows` and `MoTableSize` functions, it could be extracted into a separate function that can be called from both places.

Overall, the changes in this pull request seem to address the identified issues and improve the code by avoiding unnecessary loading of object metadata. However, there are some areas that could be improved for clarity and maintainability.

**Here are review comments for file pkg/vm/engine/disttae/cache/catalog.go:**
Title: sync_to_1.1: avoiding load object meta when object stats are valid in `mo_table_size` and `metadata_scan`

Body: This pull request addresses two issues: #13013 and #13151. It makes the following changes to the `catalog.go` file in the `pkg/vm/engine/disttae/cache` directory:

1. Adds a debug log to dump catalog cache when `ExpectedEOB` for issue #13151.
2. Avoids loading object meta when object stats are valid in the `mo_table_size` and `metadata_scan` functions.

Additionally, the pull request adds some debug functions and types for debugging purposes. These functions allow traversing the database and table caches and return debug information about the items in the cache.

Review:

1. The addition of the debug log for dumping the catalog cache is useful for debugging issue #13151. However, it is important to ensure that this debug log is removed before merging the code into the main branch to avoid unnecessary logging in production.

2. The change to avoid loading object meta when object stats are valid in the `mo_table_size` and `metadata_scan` functions is a good improvement. It can help optimize the performance of these functions by avoiding unnecessary loading of object meta when it is not needed.

3. The addition of the debug functions and types can be useful for debugging purposes. However, it is important to ensure that these debug functions are not exposed or used in production code. They should be removed or properly guarded with debug flags before merging the code into the main branch.

Overall, this pull request addresses the mentioned issues and makes some useful improvements. However, it is important to address the mentioned concerns and ensure that the debug code is properly handled before merging the changes.

**Here are review comments for file pkg/vm/engine/disttae/engine.go:**
Review:

Title: sync_to_1.1: avoiding load object meta when object stats are valid in `mo_table_size` and `metadata_scan`

The title of the pull request is clear and concise, indicating that the changes made in the pull request are related to avoiding loading object meta when object stats are valid in `mo_table_size` and `metadata_scan`.

Body:
The body of the pull request provides some additional context and information about the changes made. It mentions that the PR fixes issues #13013 and #13151. It also explains that the changes include adding debug logs to dump catalog cache and avoiding loading object meta when object stats are valid in `mo_table_size` and `metadata_scan`.

Changes:
The changes made in the pull request are focused on the file `pkg/vm/engine/disttae/engine.go`. Here are the notable changes:

1. Import statements:
   - The import statement for `bytes` package has been added.
   - The import statement for `fmt` package has been added.
   - The import statement for `sort` package has been added.
   - The import statement for `github.com/matrixorigin/matrixone/pkg/sql/plan/function` package has been added.

2. Function `enableDebug()` has been added:
   - This function enables debug mode for issue #13151.
   - It sets the `DebugGetDatabaseExpectedEOB` function from the `function` package to a custom implementation in the `Engine` struct.

3. Function `DebugGetDatabaseExpectedEOB()` has been added:
   - This function is used to debug issue #13151.
   - It traverses the databases and tables in the catalog and prints their information.
   - It also logs the current transaction state.

4. Function `Create()` has been modified:
   - The `Create()` function now uses the `getTransaction()` method to retrieve the transaction.
   - If the transaction is `nil`, it returns an error.

Security Issues:
Based on the provided information, there don't appear to be any security issues in the pull request.

Suggestions for Optimization:
1. The added import statements for `bytes`, `fmt`, and `sort` packages are not used in the code. It is recommended to remove these unused import statements to keep the code clean and avoid confusion.

2. The added import statement for `github.com/matrixorigin/matrixone/pkg/sql/plan/function` package is used in the code. However, it is not clear from the provided information whether this package is necessary for the functionality of the code. If it is not necessary, it is recommended to remove the import statement to reduce unnecessary dependencies.

3. The `enableDebug()` and `DebugGetDatabaseExpectedEOB()` functions seem to be specific to debugging issue #13151. It is recommended to add comments explaining the purpose and usage of these functions to improve code readability and maintainability.

4. The `DebugGetDatabaseExpectedEOB()` function could be optimized by using a `strings.Builder` instead of a `bytes.Buffer` for concatenating strings. This can improve performance and reduce memory allocations.

5. It is recommended to add error handling for the `Create()` function. Currently, if the transaction is `nil`, the function panics. It would be better to return an error instead of panicking to provide better error handling and prevent unexpected crashes.

Overall, the changes made in the pull request seem to address the mentioned issues and improve the code. However, the suggestions for optimization and error handling should be considered to further enhance the codebase.

**Here are review comments for file pkg/vm/engine/disttae/logtailreplay/partition_state.go:**
Review:

Title:
The title of the pull request, "sync_to_1.1: avoiding load object meta when object stats are valid in `mo_table_size` and `metadata_scan`," is clear and concise. It indicates that the changes made in this pull request are related to avoiding loading object meta when object stats are valid in the `mo_table_size` and `metadata_scan` functions.

Body:
The body of the pull request provides some context and information about the changes made. It mentions that the PR fixes issues #13013 and #13151. It also explains that the changes include adding a debug log to dump catalog cache for issue #13151 and avoiding loading object meta when object stats are valid in the `mo_table_size` and `metadata_scan` functions.

Changes:
The changes made in the pull request are focused on the file `pkg/vm/engine/disttae/logtailreplay/partition_state.go`. The changes include adding a new method `StatsValid()` to the `ObjectInfo` struct and implementing it to check if the object stats are valid. This method returns a boolean value indicating whether the object stats are valid or not.

Suggestions:
1. Security: It is important to ensure that the `StatsValid()` method is implemented correctly and does not introduce any security vulnerabilities. It would be beneficial to review the implementation and consider any potential security risks, such as input validation or potential data leakage.

2. Optimization: The changes made in this pull request seem to be focused on improving performance by avoiding unnecessary loading of object meta when object stats are already valid. It would be helpful to provide some performance metrics or benchmarks to demonstrate the impact of these changes. Additionally, it would be beneficial to consider any other potential optimizations that could be made in this code.

3. Code readability: The changes made in this pull request are relatively small and focused. However, it would be helpful to provide more comments or documentation to explain the purpose and functionality of the `StatsValid()` method. This would make the code more readable and easier to understand for future maintainers.

Overall, the changes made in this pull request seem to address the stated issues and improve performance. However, it is important to carefully review the implementation of the `StatsValid()` method for any potential security vulnerabilities. Additionally, providing more comments or documentation would enhance the code readability.

**Here are review comments for file pkg/vm/engine/disttae/txn_table.go:**
Review:

Title: sync_to_1.1: avoiding load object meta when object stats are valid in `mo_table_size` and `metadata_scan`

Body: 
- This pull request fixes issues #13013 and #13151.
- The changes made in this pull request are:
  1. Adding debug log to dump catalog cache when `ExpectedEOB` for issue #13151.
  2. Avoid loading object meta when object stats are valid in `mo_table_size` and `metadata_scan`.

Changes in `pkg/vm/engine/disttae/txn_table.go`:
- In the `Size` function, a check is added to avoid loading object meta when object stats are valid. This is done by checking if `obj.StatsValid()` is true, and if so, the size of the object is added to the `ret` variable.
- In the `GetColumMetadataScanInfo` function, a similar check is added to avoid loading object meta when object stats are valid. If `name` is equal to "AllColumns" and `obj.StatsValid()` is true, the necessary information is added to the `infoList` without loading the object meta.

Overall, the changes in this pull request aim to optimize the code by avoiding unnecessary loading of object meta when object stats are already valid.

Suggestions for improvement:
- The changes made in this pull request seem reasonable and address the issues mentioned in the body. However, there are a few suggestions for improvement:
  1. It would be helpful to provide more context in the body of the pull request. For example, explaining why loading object meta is avoided when object stats are valid and how it improves performance or addresses the mentioned issues.
  2. Consider adding comments in the code to explain the purpose of the added checks and the optimizations made.
  3. It would be beneficial to include test cases to ensure that the changes do not introduce any regressions and that the optimizations work as expected.

Security concerns:
- Based on the provided information, there are no apparent security concerns in the changes made in this pull request. However, a more thorough security review may be necessary to ensure that the optimizations do not introduce any vulnerabilities or security risks.

**Here are review comments for file test/distributed/cases/function/table_func_metadata_scan.sql:**
Title: sync_to_1.1: avoiding load object meta when object stats are valid in `mo_table_size` and `metadata_scan`

Body: This pull request addresses two issues (#13013 and #13151) by making improvements to the `mo_table_size` and `metadata_scan` functions. The changes include adding a debug log to dump catalog cache for issue #13151 and avoiding loading object metadata when object stats are valid.

Changes for file test/distributed/cases/function/table_func_metadata_scan.sql:
- Line 22: Added a new test case that selects the count of rows from the `metadata_scan` function with the table name and column name specified.
- Line 25-26: Added two new test cases that select the column name, rows count, null count, and origin size from the `metadata_scan` function with the table name and column name specified. The first test case is ignored for lines 2 and 3, and the second test case is ignored for line 0.
- Line 28: Added a new test case that selects the sum of the origin size from the `metadata_scan` function with the table name and column name specified.
- Line 30-31: Added two new test cases that select the minimum and maximum values of a column, and the sum of another column, from the `metadata_scan` function with the table name and column name specified.

Overall, the changes in this pull request aim to improve the performance and functionality of the `mo_table_size` and `metadata_scan` functions.

Upon reviewing the pull request, I have identified the following issues and suggestions for improvement:

1. In the pull request body, it would be helpful to provide more context about the issues being addressed. Specifically, what are the problems with the `mo_table_size` and `metadata_scan` functions that these changes aim to solve? Providing a brief explanation of the issues and their impact would help reviewers understand the significance of the changes.

2. The title of the pull request could be more descriptive. Instead of "sync_to_1.1: avoiding load object meta when object stats are valid in `mo_table_size` and `metadata_scan`", a more informative title could be "Improving performance and functionality of `mo_table_size` and `metadata_scan` functions".

3. In the changes made to the file `test/distributed/cases/function/table_func_metadata_scan.sql`, it is unclear why the new test cases are being added and what specific scenarios they are testing. Adding comments or explanations for each test case would make it easier for reviewers to understand the purpose of the tests.

4. The use of `@ignore` directives in the test cases is not explained in the pull request. It would be helpful to provide a brief explanation of why these directives are necessary and what they achieve.

5. It is important to ensure that the changes made in this pull request do not introduce any security vulnerabilities. Reviewers should carefully analyze the code changes to identify any potential security issues and suggest appropriate mitigation strategies if necessary.

To optimize the changes made in this pull request, the following suggestions can be considered:

1. Consider providing more detailed documentation or comments in the code to explain the purpose and functionality of the `mo_table_size` and `metadata_scan` functions. This would make it easier for future developers to understand and maintain the code.

2. If possible, provide performance benchmarks or metrics to demonstrate the improvements achieved by the changes made in this pull request. This would help validate the effectiveness of the optimizations.

3. Consider conducting thorough testing to ensure that the changes do not introduce any regressions or compatibility issues with existing functionality. This could include running the existing test suite and adding new test cases to cover different scenarios.

Overall, this pull request shows promise in improving the performance and functionality of the `mo_table_size` and `metadata_scan` functions. However, addressing the identified issues and implementing the suggested improvements would further enhance the quality and maintainability of the codebase.

**Review by m-schen — APPROVED** [review]

**Review by aressu1985 — APPROVED** [review]

**Review by reusee — APPROVED** [review]

**Review by XuPeng-SH — APPROVED** [review]

**Review by sukki37 — APPROVED** [review]

**Pull request merged**
